blog/_src/posts/2018-08-16-if-you-give-a-mouse-a-macro.md

```diff
@@ -0,0 +1,445 @@
+a procedure argument, `let`, or `define`; alternatively, we can provide syntax
+that evaluates to a function, establishing a transformer binding invoked on
+references to the identifier. Using both modes, we can define a function that
+sets up a Turnstile-style type rename:
+
+```racket
+(define-for-syntax (int-def-ctx-bind-type-rename x x- t ctx)
+  (syntax-local-bind-syntaxes (list x)
+                              #`(make-rename-transformer
+                                 (add-orig (assign-type #'#,x- #'#,t #:wrap? #f) #'#,x))
+                              ctx)
+  (syntax-local-bind-syntaxes (list x-) #f ctx))
+```
+
+The first call binds `x` to a transformer that renames to `x-`; the second lets
+the expander know that we are taking care of making sure that `x-` will actually
+be bound to something.
+
+Our `define` form must communicate the information needed to call
+`int-def-ctx-bind-type-rename` back out to the surrounding context. One way to
+do this is to add an intermediate step to the expansion of `define` that
+includes the necessary information as part of its syntax. Then, the surrounding
+context can analyze the expansion of each term, looking for that form.
+
+Concretely, `define` will expand to `define/intermediate`, which will in turn
+expand to what `define` originally expanded to:
+
+```racket
+(define-typed-syntax define
+  [(_ x:id e) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   #:with x- (generate-temporary #'x)
+   #:with x+ (syntax-local-identifier-as-binding #'x)
+   -----------------------------------------------------
+   [⊢ (define/intermediate x+ x- τ e-) ⇒ Void]]
+  [(_ (f [x (~datum :) τ] ...) e ...+) ≫
+   -----------------------------------
+   [≻ (define f (λ+ ([x : τ] ...) e ...))]])         
+
+(define-syntax (define/intermediate stx)
+  (syntax-parse stx
+    [(_ x:id x-:id τ e)
+     #'(begin-
+         (define-typed-variable-rename x ≫ x- : τ)
+         (define- x- e)
+         a-deep-dark-void)]))
+```
+
+The reason we create an `x+` using [`syntax-local-identifier-as-binding`][13] is
+[due to a bug in the expander][12]. The explanation is rather involved and
+frankly I only barely understand what's going on myself (if at all), so let's
+just leave it at that and move on.
+
+Then, for each form `e` in a sequence, we can call `local-expand` with `def-ctx`
+and then check the expansion, `e-`, for `define/intermediate`. In those cases,
+we can use `int-def-ctx-bind-type-rename` to add it to the context. The
+procedure `add-bindings-to-ctx` performs this check on an expanded form `e-`:
+
+```racket
+(define-for-syntax (add-bindings-to-ctx e- def-ctx)
+  (syntax-parse e-
+        #:literals (erased define/intermediate)
+        [(erased (define/intermediate x:id x-:id τ e-))
+         (int-def-ctx-bind-type-rename #'x #'x- #'τ def-ctx)]
+        [_ (void)]))
+        
+```
+
+One extra detail that snuck in to the implementation is the presence of the
```

Comment:
I guess the important thing isn't `erased`, but that you can use the stop list parameter to `local-expand` to look at the output of Turnstile rules (or any other macro). So the surprising thing is that we need to use `erased` in the stop list, in order to play nicely with Turnstile, rather than `define-intermediate`. I'll try adding an explanation for this.